In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('data/survey_results_public.csv')

In [4]:
df.head(2)

Respondent                              MainBranch Hobbyist  \
0           1  I am a student who is learning to code      Yes   
1           2  I am a student who is learning to code       No   

               OpenSourcer                                         OpenSource  \
0                    Never  The quality of OSS and closed source software ...   
1  Less than once per year  The quality of OSS and closed source software ...   

                               Employment                 Country  \
0  Not employed, and not looking for work          United Kingdom   
1      Not employed, but looking for work  Bosnia and Herzegovina   

          Student                                            EdLevel  \
0              No                          Primary/elementary school   
1  Yes, full-time  Secondary school (e.g. American high school, G...   

  UndergradMajor  ...                            WelcomeChange  \
0            NaN  ...  Just as welcome now as I felt last year   
1            NaN  ...  Just as welcome now as I felt last year   

                                        SONewContent   Age Gender Trans  \
0  Tech articles written by other developers;Indu...  14.0    Man    No   
1  Tech articles written by other developers;Indu...  19.0    Man    No   

                 Sexuality Ethnicity Dependents           SurveyLength  \
0  Straight / Heterosexual       NaN         No  Appropriate in length   
1  Straight / Heterosexual       NaN         No  Appropriate in length   

                   SurveyEase  
0  Neither easy nor difficult  
1  Neither easy nor difficult  

[2 rows x 85 columns]

In [5]:
df.shape

(88883, 85)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88883 entries, 0 to 88882
Data columns (total 85 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Respondent              88883 non-null  int64  
 1   MainBranch              88331 non-null  object 
 2   Hobbyist                88883 non-null  object 
 3   OpenSourcer             88883 non-null  object 
 4   OpenSource              86842 non-null  object 
 5   Employment              87181 non-null  object 
 6   Country                 88751 non-null  object 
 7   Student                 87014 non-null  object 
 8   EdLevel                 86390 non-null  object 
 9   UndergradMajor          75614 non-null  object 
 10  EduOther                84260 non-null  object 
 11  OrgSize                 71791 non-null  object 
 12  DevType                 81335 non-null  object 
 13  YearsCode               87938 non-null  object 
 14  Age1stCode              87634 non-null

### To view all columns in dataframe

In [7]:
pd.set_option('display.max_columns', 85)

In [8]:
df.head(3)

Respondent                                         MainBranch Hobbyist  \
0           1             I am a student who is learning to code      Yes   
1           2             I am a student who is learning to code       No   
2           3  I am not primarily a developer, but I write co...      Yes   

               OpenSourcer                                         OpenSource  \
0                    Never  The quality of OSS and closed source software ...   
1  Less than once per year  The quality of OSS and closed source software ...   
2                    Never  The quality of OSS and closed source software ...   

                               Employment                 Country  \
0  Not employed, and not looking for work          United Kingdom   
1      Not employed, but looking for work  Bosnia and Herzegovina   
2                      Employed full-time                Thailand   

          Student                                            EdLevel  \
0              No                          Primary/elementary school   
1  Yes, full-time  Secondary school (e.g. American high school, G...   
2              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                  UndergradMajor  \
0                            NaN   
1                            NaN   
2  Web development or web design   

                                            EduOther               OrgSize  \
0  Taught yourself a new language, framework, or ...                   NaN   
1  Taken an online course in programming or softw...                   NaN   
2  Taught yourself a new language, framework, or ...  100 to 499 employees   

                                             DevType YearsCode Age1stCode  \
0                                                NaN         4         10   
1  Developer, desktop or enterprise applications;...       NaN         17   
2  Designer;Developer, back-end;Developer, front-...         3         22   

  YearsCodePro           CareerSat              JobSat              MgrIdiot  \
0          NaN                 NaN                 NaN                   NaN   
1          NaN                 NaN                 NaN                   NaN   
2            1  Slightly satisfied  Slightly satisfied  Not at all confident   

   MgrMoney   MgrWant                                            JobSeek  \
0       NaN       NaN                                                NaN   
1       NaN       NaN                    I am actively looking for a job   
2  Not sure  Not sure  I’m not actively looking, but I am open to new...   

           LastHireDate                              LastInt FizzBuzz  \
0                   NaN                                  NaN      NaN   
1  I've never had a job                                  NaN      NaN   
2         1-2 years ago  Interview with people in peer roles       No   

                                          JobFactors  \
0                                                NaN   
1  Financial performance or funding status of the...   
2  Languages, frameworks, and other technologies ...   

                                        ResumeUpdate CurrencySymbol  \
0                                                NaN            NaN   
1  Something else changed (education, award, medi...            NaN   
2                   I was preparing for a job search            THB   

  CurrencyDesc  CompTotal CompFreq  ConvertedComp  WorkWeekHrs  \
0          NaN        NaN      NaN            NaN          NaN   
1          NaN        NaN      NaN            NaN          NaN   
2    Thai baht    23000.0  Monthly         8820.0         40.0   

                                            WorkPlan  \
0                                                NaN   
1                                                NaN   
2  There's no schedule or spec; I work on what se...   

                                       WorkChallenge  \
0                                                NaN   
1    

### As we see many columns and want to know hte columns desctription we can load other data that consists columns details

In [9]:
schema_df = pd.read_csv('data/survey_results_schema.csv')

In [10]:
schema_df

Column                                       QuestionText
0     Respondent  Randomized respondent ID number (not in order ...
1     MainBranch  Which of the following options best describes ...
2       Hobbyist                            Do you code as a hobby?
3    OpenSourcer        How often do you contribute to open source?
4     OpenSource  How do you feel about the quality of open sour...
..           ...                                                ...
80     Sexuality  Which of the following do you currently identi...
81     Ethnicity  Which of the following do you identify as? Ple...
82    Dependents  Do you have any dependents (e.g., children, el...
83  SurveyLength  How do you feel about the length of the survey...
84    SurveyEase  How easy or difficult was this survey to compl...

[85 rows x 2 columns]

In [11]:
pd.set_option('display.max_rows', 85)

In [12]:
schema_df

Column                                       QuestionText
0               Respondent  Randomized respondent ID number (not in order ...
1               MainBranch  Which of the following options best describes ...
2                 Hobbyist                            Do you code as a hobby?
3              OpenSourcer        How often do you contribute to open source?
4               OpenSource  How do you feel about the quality of open sour...
5               Employment  Which of the following best describes your cur...
6                  Country          In which country do you currently reside?
7                  Student  Are you currently enrolled in a formal, degree...
8                  EdLevel  Which of the following best describes the high...
9           UndergradMajor  What was your main or most important field of ...
10                EduOther  Which of the following types of non-degree edu...
11                 OrgSize  Approximately how many people are employed by ...
12                 DevType  Which of the following describe you? Please se...
13               YearsCode  Including any education, how many years have y...
14              Age1stCode  At what age did you write your first line of c...
15            YearsCodePro  How many years have you coded professionally (...
16               CareerSat  Overall, how satisfied are you with your caree...
17                  JobSat  How satisfied are you with your current job? (...
18                MgrIdiot  How confident are you that your manager knows ...
19                MgrMoney  Do you believe that you need to be a manager t...
20                 MgrWant  Do you want to become a manager yourself in th...
21                 JobSeek  Which of the following best describes your cur...
22            LastHireDate  When was the last time that you took a job wit...
23                 LastInt  In your most recent successful job interview (...
24                FizzBuzz  Have you ever been asked to solve FizzBuzz in ...
25              JobFactors  Imagine that you are deciding between two job ...
26            ResumeUpdate  Think back to the last time you updated your r...
27          CurrencySymbol  Which currency do you use day-to-day? If your ...
28            CurrencyDesc  Which currency do you use day-to-day? If your ...
29               CompTotal  What is your current total compensation (salar...
30                CompFreq   Is that compensation weekly, monthly, or yearly?
31           ConvertedComp  Salary converted to annual USD salaries using ...
32             WorkWeekHrs   On average, how many hours per week do you work?
33                WorkPlan            How structured or planned is your work?
34           WorkChallenge  Of these options, what are your greatest chall...
35              WorkRemote                    How often do you work remotely?
36                 WorkLoc                    Where would you prefer to work?
37                  ImpSyn  For the specific work you do, and the years of...
38                 CodeRev           Do you review code as part of your work?
39              CodeRevHrs  On average, how many hours per week do you spe...
40               UnitTests  Does your company regularly employ unit tests ...
41             PurchaseHow  How does your company make decisions about pur...
42            PurchaseWhat  What level of influence do you, personally, ha...
43      LanguageWorkedWith  Which of the following programming, scripting,...
44  LanguageDesireNextYear  Which of the following programming, scripting,...
45      DatabaseWorkedWith  Which of the following database environments h...
46  DatabaseDesireNextYear  Which of the following database environments h...
47      PlatformWorkedWith  Which of the following platforms have you done...
48  PlatformDesireNextYear  Which of the following platforms have you done...
49      WebFrameWorkedWith  Which of the following web frameworks have you...
50  WebFrameDesireNextYear  Which of the f

In [13]:
df.columns

Index(['Respondent', 'MainBranch', 'Hobbyist', 'OpenSourcer', 'OpenSource',
       'Employment', 'Country', 'Student', 'EdLevel', 'UndergradMajor',
       'EduOther', 'OrgSize', 'DevType', 'YearsCode', 'Age1stCode',
       'YearsCodePro', 'CareerSat', 'JobSat', 'MgrIdiot', 'MgrMoney',
       'MgrWant', 'JobSeek', 'LastHireDate', 'LastInt', 'FizzBuzz',
       'JobFactors', 'ResumeUpdate', 'CurrencySymbol', 'CurrencyDesc',
       'CompTotal', 'CompFreq', 'ConvertedComp', 'WorkWeekHrs', 'WorkPlan',
       'WorkChallenge', 'WorkRemote', 'WorkLoc', 'ImpSyn', 'CodeRev',
       'CodeRevHrs', 'UnitTests', 'PurchaseHow', 'PurchaseWhat',
       'LanguageWorkedWith', 'LanguageDesireNextYear', 'DatabaseWorkedWith',
       'DatabaseDesireNextYear', 'PlatformWorkedWith',
       'PlatformDesireNextYear', 'WebFrameWorkedWith',
       'WebFrameDesireNextYear', 'MiscTechWorkedWith',
       'MiscTechDesireNextYear', 'DevEnviron', 'OpSys', 'Containers',
       'BlockchainOrg', 'BlockchainIs', 'BetterLife'

In [14]:
df['Hobbyist'].value_counts()

Yes    71257
No     17626
Name: Hobbyist, dtype: int64

In [15]:
 df.loc[0:2, 'Hobbyist':'Employment' ]

Hobbyist              OpenSourcer  \
0      Yes                    Never   
1       No  Less than once per year   
2      Yes                    Never   

                                          OpenSource  \
0  The quality of OSS and closed source software ...   
1  The quality of OSS and closed source software ...   
2  The quality of OSS and closed source software ...   

                               Employment  
0  Not employed, and not looking for work  
1      Not employed, but looking for work  
2                      Employed full-time

### Setting index

In [16]:
df.set_index('Respondent', inplace=True)

In [17]:
df.head(2)

MainBranch Hobbyist  \
Respondent                                                    
1           I am a student who is learning to code      Yes   
2           I am a student who is learning to code       No   

                        OpenSourcer  \
Respondent                            
1                             Never   
2           Less than once per year   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   

           UndergradMajor                                           EduOther  \
Respondent                                                                     
1                     NaN  Taught yourself a new language, framework, or ...   
2                     NaN  Taken an online course in programming or softw...   

           OrgSize                                            DevType  \
Respondent                                                              
1              NaN                                                NaN   
2              NaN  Developer, desktop or enterprise applications;...   

           YearsCode Age1stCode YearsCodePro CareerSat JobSat MgrIdiot  \
Respondent                                                               
1                  4         10          NaN       NaN    NaN      NaN   
2                NaN         17          NaN       NaN    NaN      NaN   

           MgrMoney MgrWant                          JobSeek  \
Respondent                                                     
1               NaN     NaN                              NaN   
2               NaN     NaN  I am actively looking for a job   

                    LastHireDate LastInt FizzBuzz  \
Respondent                                          
1                            NaN     NaN      NaN   
2           I've never had a job     NaN      NaN   

                                                   JobFactors  \
Respondent                                                      
1                                                         NaN   
2           Financial performance or funding status of the...   

                                                 ResumeUpdate CurrencySymbol  \
Respondent                                                                     
1                                                         NaN            NaN   
2           Something else changed (education, award, medi...            NaN   

           CurrencyDesc  CompTotal CompFreq  ConvertedComp  WorkWeekHrs  \
Respondent                                                                
1                   NaN        NaN      NaN            NaN          NaN   
2                   NaN        NaN      NaN            NaN          NaN   

           WorkPlan WorkChallenge WorkRemote WorkLoc ImpSyn CodeRev  \
Respondent                                                            
1               NaN           NaN        NaN     NaN    NaN     NaN   
2               NaN           NaN        NaN     NaN    NaN     NaN   

            CodeRevHrs UnitTests PurchaseHow PurchaseWhat  \
Respondent                                                  
1                  NaN       NaN         NaN          NaN   

### Resetting Index

In [18]:
df.reset_index(inplace=True)

In [19]:
df.head(2)

Respondent                              MainBranch Hobbyist  \
0           1  I am a student who is learning to code      Yes   
1           2  I am a student who is learning to code       No   

               OpenSourcer                                         OpenSource  \
0                    Never  The quality of OSS and closed source software ...   
1  Less than once per year  The quality of OSS and closed source software ...   

                               Employment                 Country  \
0  Not employed, and not looking for work          United Kingdom   
1      Not employed, but looking for work  Bosnia and Herzegovina   

          Student                                            EdLevel  \
0              No                          Primary/elementary school   
1  Yes, full-time  Secondary school (e.g. American high school, G...   

  UndergradMajor                                           EduOther OrgSize  \
0            NaN  Taught yourself a new language, framework, or ...     NaN   
1            NaN  Taken an online course in programming or softw...     NaN   

                                             DevType YearsCode Age1stCode  \
0                                                NaN         4         10   
1  Developer, desktop or enterprise applications;...       NaN         17   

  YearsCodePro CareerSat JobSat MgrIdiot MgrMoney MgrWant  \
0          NaN       NaN    NaN      NaN      NaN     NaN   
1          NaN       NaN    NaN      NaN      NaN     NaN   

                           JobSeek          LastHireDate LastInt FizzBuzz  \
0                              NaN                   NaN     NaN      NaN   
1  I am actively looking for a job  I've never had a job     NaN      NaN   

                                          JobFactors  \
0                                                NaN   
1  Financial performance or funding status of the...   

                                        ResumeUpdate CurrencySymbol  \
0                                                NaN            NaN   
1  Something else changed (education, award, medi...            NaN   

  CurrencyDesc  CompTotal CompFreq  ConvertedComp  WorkWeekHrs WorkPlan  \
0          NaN        NaN      NaN            NaN          NaN      NaN   
1          NaN        NaN      NaN            NaN          NaN      NaN   

  WorkChallenge WorkRemote WorkLoc ImpSyn CodeRev  CodeRevHrs UnitTests  \
0           NaN        NaN     NaN    NaN     NaN         NaN       NaN   
1           NaN        NaN     NaN    NaN     NaN         NaN       NaN   

  PurchaseHow PurchaseWhat               LanguageWorkedWith  \
0         NaN          NaN  HTML/CSS;Java;JavaScript;Python   
1         NaN          NaN              C++;HTML/CSS;Python   

                            LanguageDesireNextYear DatabaseWorkedWith  \
0  C;C++;C#;Go;HTML/CSS;Java;JavaScript;Python;SQL             SQLite   
1                      C++;HTML/CSS;JavaScript;SQL                NaN   

  DatabaseDesireNextYear PlatformWorkedWith   PlatformDesireNextYear  \
0                  MySQL      MacOS;Windows  Android;Arduino;Windows   
1                  MySQL            Windows                  Windows   

  WebFrameWorkedWith WebFrameDesireNextYear MiscTechWorkedWith  \
0       Django;Flask           Flask;jQuery            Node.js   
1             Django                 Django                NaN   

  MiscTechDesireNextYear                  DevEnviron    OpSys  \
0                Node.js  IntelliJ;Notepad++;PyCharm  Windows   
1                    NaN                Atom;PyCharm  Windows   

                Containers BlockchainOrg  \
0  I do not use containers           NaN   
1  I do not use containers           NaN   

                                        BlockchainIs BetterLife  \
0                                                NaN        Yes   
1  Useful across many domains and could change ma...        Yes   

                                   ITperson OffOn SocialM

### Setting Index while loading csv file

In [20]:
df = pd.read_csv('data/survey_results_public.csv', index_col='Respondent')
schema_df = pd.read_csv('data/survey_results_schema.csv', index_col='Column')

In [21]:
df.head(2)

MainBranch Hobbyist  \
Respondent                                                    
1           I am a student who is learning to code      Yes   
2           I am a student who is learning to code       No   

                        OpenSourcer  \
Respondent                            
1                             Never   
2           Less than once per year   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   

           UndergradMajor                                           EduOther  \
Respondent                                                                     
1                     NaN  Taught yourself a new language, framework, or ...   
2                     NaN  Taken an online course in programming or softw...   

           OrgSize                                            DevType  \
Respondent                                                              
1              NaN                                                NaN   
2              NaN  Developer, desktop or enterprise applications;...   

           YearsCode Age1stCode YearsCodePro CareerSat JobSat MgrIdiot  \
Respondent                                                               
1                  4         10          NaN       NaN    NaN      NaN   
2                NaN         17          NaN       NaN    NaN      NaN   

           MgrMoney MgrWant                          JobSeek  \
Respondent                                                     
1               NaN     NaN                              NaN   
2               NaN     NaN  I am actively looking for a job   

                    LastHireDate LastInt FizzBuzz  \
Respondent                                          
1                            NaN     NaN      NaN   
2           I've never had a job     NaN      NaN   

                                                   JobFactors  \
Respondent                                                      
1                                                         NaN   
2           Financial performance or funding status of the...   

                                                 ResumeUpdate CurrencySymbol  \
Respondent                                                                     
1                                                         NaN            NaN   
2           Something else changed (education, award, medi...            NaN   

           CurrencyDesc  CompTotal CompFreq  ConvertedComp  WorkWeekHrs  \
Respondent                                                                
1                   NaN        NaN      NaN            NaN          NaN   
2                   NaN        NaN      NaN            NaN          NaN   

           WorkPlan WorkChallenge WorkRemote WorkLoc ImpSyn CodeRev  \
Respondent                                                            
1               NaN           NaN        NaN     NaN    NaN     NaN   
2               NaN           NaN        NaN     NaN    NaN     NaN   

            CodeRevHrs UnitTests PurchaseHow PurchaseWhat  \
Respondent                                                  
1                  NaN       NaN         NaN          NaN   

In [23]:
schema_df.head(5)

QuestionText
Column                                                        
Respondent   Randomized respondent ID number (not in order ...
MainBranch   Which of the following options best describes ...
Hobbyist                               Do you code as a hobby?
OpenSourcer        How often do you contribute to open source?
OpenSource   How do you feel about the quality of open sour...

In [26]:
schema_df.loc['MgrMoney', 'QuestionText']  ## passing 'QuestionText' bcz otherwise sentences is incomplete

'Do you believe that you need to be a manager to make more money?'

In [28]:
schema_df.sort_index(ascending=False)

QuestionText
Column                                                                   
YearsCodePro            How many years have you coded professionally (...
YearsCode               Including any education, how many years have y...
WorkWeekHrs              On average, how many hours per week do you work?
WorkRemote                                How often do you work remotely?
WorkPlan                          How structured or planned is your work?
WorkLoc                                   Where would you prefer to work?
WorkChallenge           Of these options, what are your greatest chall...
WelcomeChange           Compared to last year, how welcome do you feel...
WebFrameWorkedWith      Which of the following web frameworks have you...
WebFrameDesireNextYear  Which of the following web frameworks have you...
UnitTests               Does your company regularly employ unit tests ...
UndergradMajor          What was your main or most important field of ...
Trans                                     Do you identify as transgender?
SurveyLength            How do you feel about the length of the survey...
SurveyEase              How easy or difficult was this survey to compl...
Student                 Are you currently enrolled in a formal, degree...
SocialMedia                   What social media site do you use the most?
Sexuality               Which of the following do you currently identi...
ScreenName                                           What do you call it?
SOVisitTo               I visit Stack Overflow to... (check all that a...
SOVisitFreq             How frequently would you say you visit Stack O...
SOVisit1st              To the best of your memory, when did you first...
SOTimeSaved             Think back to the last time you solved a codin...
SOPartFreq              How frequently would you say you participate i...
SONewContent            Would you like to see any of the following on ...
SOJobs                  Have you ever used or visited Stack Overflow J...
SOHowMuchTime           About how much time did you save? If you're no...
SOFindAnswer            On average, how many times a week do you find ...
SOComm                  Do you consider yourself a member of the Stack...
SOAccount                           Do you have a Stack Overflow account?
ResumeUpdate            Think back to the last time you updated your r...
Respondent              Randomized respondent ID number (not in order ...
PurchaseWhat            What level of influence do you, personally, ha...
PurchaseHow             How does your company make decisions about pur...
PlatformWorkedWith      Which of the following platforms have you done...
PlatformDesireNextYear  Which of the following platforms have you done...
OrgSize                 Approximately how many people are employed by ...
OpenSourcer                   How often do you contribute to open source?
OpenSource              How do you feel about the quality of open sour...
OpSys                   What is the primary operating system in which ...
OffOn                         Have you tried turning it off and on again?
MiscTechWorkedWith      Which of the following other frameworks, libra...
MiscTechDesireNextYear  Which of the following other frameworks, libra...
MgrWant                 Do you want to become a manager yourself in th...
MgrMoney                Do you believe that you need to be a manager t...
MgrIdiot                How confident are you that your manager knows ...
MainBranch              Which of the following options best describes ...
LastInt                 In your most recent successful job interview (...
LastHireDate            When was the last time that you took a job wit...
LanguageWorkedWith      Which of the following programming, scripting,...
LanguageDesireNextYear  Which of the following programming, scripting,...
JobSeek                 Which of the following best describes your cur...
JobSat                  How satisfied are you with your current j

### Filtering data on different conditions

In [32]:
high_salary = df['ConvertedComp'] > 80000

In [33]:
df.loc[high_salary, ['Country','LanguageWorkedWith','ConvertedComp']]

Country                                 LanguageWorkedWith  \
Respondent                                                                      
6                   Canada                                         Java;R;SQL   
9              New Zealand  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;P...   
13           United States  Bash/Shell/PowerShell;HTML/CSS;JavaScript;PHP;...   
16          United Kingdom  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;T...   
22           United States  Bash/Shell/PowerShell;C++;HTML/CSS;JavaScript;...   
...                    ...                                                ...   
88876        United States  Bash/Shell/PowerShell;C#;HTML/CSS;Java;Python;...   
88877        United States  Bash/Shell/PowerShell;C;Clojure;HTML/CSS;Java;...   
88878        United States               HTML/CSS;JavaScript;Scala;TypeScript   
88879              Finland                   Bash/Shell/PowerShell;C++;Python   
88882          Netherlands             C#;HTML/CSS;Java;JavaScript;PHP;Python   

            ConvertedComp  
Respondent                 
6                366420.0  
9                 95179.0  
13                90000.0  
16               455352.0  
22               103000.0  
...                   ...  
88876            180000.0  
88877           2000000.0  
88878            130000.0  
88879             82488.0  
88882            588012.0  

[18944 rows x 3 columns]

In [34]:
countries = ['Canada', 'New Zeland', 'United States', 'India']
filter_data = df['Country'].isin(countries)

In [35]:
filter_data

Respondent
1        False
2        False
3        False
4         True
5        False
         ...  
88377     True
88601    False
88802    False
88816    False
88863    False
Name: Country, Length: 88883, dtype: bool

In [36]:
df.loc[filter_data, 'Country']

Respondent
4        United States
6               Canada
8                India
10               India
12              Canada
             ...      
85182           Canada
85642    United States
86012            India
88282    United States
88377           Canada
Name: Country, Length: 33405, dtype: object

### To see people working with python only

In [37]:
filter_data_python = df['LanguageWorkedWith'].str.contains('Python', na=False)

In [38]:
df.loc[filter_data_python, 'LanguageWorkedWith']

Respondent
1                          HTML/CSS;Java;JavaScript;Python
2                                      C++;HTML/CSS;Python
4                                      C;C++;C#;Python;SQL
5              C++;HTML/CSS;Java;JavaScript;Python;SQL;VBA
8        Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
                               ...                        
84539    Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
85738      Bash/Shell/PowerShell;C++;Python;Ruby;Other(s):
86566      Bash/Shell/PowerShell;HTML/CSS;Python;Other(s):
87739             C;C++;HTML/CSS;JavaScript;PHP;Python;SQL
88212                           HTML/CSS;JavaScript;Python
Name: LanguageWorkedWith, Length: 36443, dtype: object

### Updating rows and columns

In [41]:
df.rename(columns={'ConvertedComp': 'Salary(USD)'}, inplace=True)

In [43]:
df['Salary(USD)']

Respondent
1            NaN
2            NaN
3         8820.0
4        61000.0
5            NaN
          ...   
88377        NaN
88601        NaN
88802        NaN
88816        NaN
88863        NaN
Name: Salary(USD), Length: 88883, dtype: float64

### map function

In [44]:
df['Hobbyist'].map({'Yes': True, 'No': False})  # No inplace== True argument in map function

Respondent
1         True
2        False
3         True
4        False
5         True
         ...  
88377     True
88601    False
88802    False
88816    False
88863     True
Name: Hobbyist, Length: 88883, dtype: bool

In [45]:
# So, to change it ike inplace = True

df['Hobbyist'] = df['Hobbyist'].map({'Yes': True, 'No': False})

In [46]:
df.head(2)

MainBranch  Hobbyist  \
Respondent                                                     
1           I am a student who is learning to code      True   
2           I am a student who is learning to code     False   

                        OpenSourcer  \
Respondent                            
1                             Never   
2           Less than once per year   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   

           UndergradMajor                                           EduOther  \
Respondent                                                                     
1                     NaN  Taught yourself a new language, framework, or ...   
2                     NaN  Taken an online course in programming or softw...   

           OrgSize                                            DevType  \
Respondent                                                              
1              NaN                                                NaN   
2              NaN  Developer, desktop or enterprise applications;...   

           YearsCode Age1stCode YearsCodePro CareerSat JobSat MgrIdiot  \
Respondent                                                               
1                  4         10          NaN       NaN    NaN      NaN   
2                NaN         17          NaN       NaN    NaN      NaN   

           MgrMoney MgrWant                          JobSeek  \
Respondent                                                     
1               NaN     NaN                              NaN   
2               NaN     NaN  I am actively looking for a job   

                    LastHireDate LastInt FizzBuzz  \
Respondent                                          
1                            NaN     NaN      NaN   
2           I've never had a job     NaN      NaN   

                                                   JobFactors  \
Respondent                                                      
1                                                         NaN   
2           Financial performance or funding status of the...   

                                                 ResumeUpdate CurrencySymbol  \
Respondent                                                                     
1                                                         NaN            NaN   
2           Something else changed (education, award, medi...            NaN   

           CurrencyDesc  CompTotal CompFreq  Salary(USD)  WorkWeekHrs  \
Respondent                                                              
1                   NaN        NaN      NaN          NaN          NaN   
2                   NaN        NaN      NaN          NaN          NaN   

           WorkPlan WorkChallenge WorkRemote WorkLoc ImpSyn CodeRev  \
Respondent                                                            
1               NaN           NaN        NaN     NaN    NaN     NaN   
2               NaN           NaN        NaN     NaN    NaN     NaN   

            CodeRevHrs UnitTests PurchaseHow PurchaseWhat  \
Respondent                                                  
1                  NaN       NaN         NaN          NaN   
2  

### Sorting Data (Multiple Values)

In [47]:
df.sort_values(by=['Country','Salary(USD)'], ascending=[True, False], inplace=True)

In [49]:
df[['Country', 'Salary(USD)']].head()

Country  Salary(USD)
Respondent                          
63129       Afghanistan    1000000.0
50499       Afghanistan     153216.0
39258       Afghanistan      19152.0
58450       Afghanistan      17556.0
7085        Afghanistan      14364.0

### nlargest() method

In [50]:
df['Salary(USD)'].nlargest(10)

Respondent
25983    2000000.0
87896    2000000.0
22013    2000000.0
28243    2000000.0
72732    2000000.0
78151    2000000.0
80200    2000000.0
52132    2000000.0
75561    2000000.0
32250    2000000.0
Name: Salary(USD), dtype: float64

In [51]:
## Seeing in dataframe
df.nlargest(10, 'Salary(USD)')

MainBranch  Hobbyist  \
Respondent                                                                
25983                          I am a developer by profession      True   
87896                          I am a developer by profession      True   
22013                          I am a developer by profession      True   
28243                          I am a developer by profession      True   
72732       I am not primarily a developer, but I write co...     False   
78151                          I am a developer by profession      True   
80200                          I am a developer by profession      True   
52132                          I am a developer by profession      True   
75561                          I am a developer by profession      True   
32250                          I am a developer by profession      True   

                                                  OpenSourcer  \
Respondent                                                      
25983                                 Less than once per year   
87896                                 Less than once per year   
22013                                                   Never   
28243                              Once a month or more often   
72732       Less than once a month but more than once per ...   
78151                                                   Never   
80200                                                   Never   
52132       Less than once a month but more than once per ...   
75561       Less than once a month but more than once per ...   
32250                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
25983       OSS is, on average, of HIGHER quality than pro...   
87896       The quality of OSS and closed source software ...   
22013       The quality of OSS and closed source software ...   
28243       OSS is, on average, of HIGHER quality than pro...   
72732       OSS is, on average, of LOWER quality than prop...   
78151       OSS is, on average, of HIGHER quality than pro...   
80200       OSS is, on average, of LOWER quality than prop...   
52132       OSS is, on average, of HIGHER quality than pro...   
75561       The quality of OSS and closed source software ...   
32250       The quality of OSS and closed source software ...   

                                                   Employment      Country  \
Respondent                                                                   
25983                                      Employed full-time       Canada   
87896                                      Employed full-time      Germany   
22013                                      Employed full-time        India   
28243       Independent contractor, freelancer, or self-em...        India   
72732                                                     NaN        India   
78151                                      Employed full-time       Mexico   
80200                                      Employed full-time  Netherlands   
52132                                      Employed full-time         Peru   
75561                                      Employed full-time    Singapore   
32250                                      Employed full-time  Switzerland   

                   Student                                            EdLevel  \
Respondent                                                                      
25983                   No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
87896                   No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
22013                   No                 Professional degree (JD, MD, etc.)   
28243                   No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
72732       Yes, full-time           Bachelor’s degree (BA, BS, B.Eng., etc.)   
78151                   No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
802

In [60]:
filt = df['Country'] == 'India'


In [61]:
df.loc[filt]['SocialMedia'].value_counts()
## Social Media Used in country Canada

WhatsApp                    2990
YouTube                     1820
LinkedIn                     955
Facebook                     841
Instagram                    822
Twitter                      542
Reddit                       473
I don't use social media     250
Snapchat                      23
WeChat 微信                      5
Hello                          5
VK ВКонта́кте                  4
Youku Tudou 优酷                 2
Weibo 新浪微博                     1
Name: SocialMedia, dtype: int64

### group by

In [70]:
country_group = df.groupby(['Country'])
country_group

In [75]:
country_group.get_group('India')

MainBranch  Hobbyist  \
Respondent                                                                
22013                          I am a developer by profession      True   
28243                          I am a developer by profession      True   
72732       I am not primarily a developer, but I write co...     False   
16793                          I am a developer by profession      True   
52561                          I am a developer by profession      True   
...                                                       ...       ...   
77339                                                     NaN      True   
79795                                                     NaN      True   
83862                                                     NaN      True   
84299                                                     NaN      True   
86012                                                     NaN      True   

                                                  OpenSourcer  \
Respondent                                                      
22013                                                   Never   
28243                              Once a month or more often   
72732       Less than once a month but more than once per ...   
16793                              Once a month or more often   
52561                                                   Never   
...                                                       ...   
77339                                 Less than once per year   
79795       Less than once a month but more than once per ...   
83862                              Once a month or more often   
84299                                                   Never   
86012                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
22013       The quality of OSS and closed source software ...   
28243       OSS is, on average, of HIGHER quality than pro...   
72732       OSS is, on average, of LOWER quality than prop...   
16793       OSS is, on average, of HIGHER quality than pro...   
52561       OSS is, on average, of LOWER quality than prop...   
...                                                       ...   
77339       OSS is, on average, of HIGHER quality than pro...   
79795       OSS is, on average, of HIGHER quality than pro...   
83862       OSS is, on average, of HIGHER quality than pro...   
84299       The quality of OSS and closed source software ...   
86012       OSS is, on average, of HIGHER quality than pro...   

                                                   Employment Country  \
Respondent                                                              
22013                                      Employed full-time   India   
28243       Independent contractor, freelancer, or self-em...   India   
72732                                                     NaN   India   
16793       Independent contractor, freelancer, or self-em...   India   
52561                                      Employed full-time   India   
...                                                       ...     ...   
77339                                      Employed full-time   India   
79795                                      Employed full-time   India   
83862                                      Employed full-time   India   
84299                                      Employed full-time   India   
86012                                      Employed full-time   India   

                   Student                                      EdLevel  \
Respondent                                                                
22013                   No           Professional degree (JD, MD, etc.)   
28243                   No  Master’s degree (MA, MS, M.Eng., MBA, etc.)   
72732       Yes, full-time     Bachelor’s degree (BA, BS, B.Eng., etc.)   
16793                   No     Bachelor’s degree (BA, BS, 

In [72]:
country_group['SocialMedia'].value_counts().head(10)

Country      SocialMedia             
Afghanistan  Facebook                    15
             YouTube                      9
             I don't use social media     6
             WhatsApp                     4
             Instagram                    1
             LinkedIn                     1
             Twitter                      1
Albania      WhatsApp                    18
             Facebook                    16
             Instagram                   13
Name: SocialMedia, dtype: int64

In [74]:
## Can also do
country_group['SocialMedia'].value_counts().loc['Canada']

SocialMedia
Reddit                      901
YouTube                     526
Facebook                    515
Twitter                     405
Instagram                   246
I don't use social media    232
WhatsApp                    185
LinkedIn                    179
WeChat 微信                    44
Snapchat                     28
VK ВКонта́кте                 7
Weibo 新浪微博                    1
Name: SocialMedia, dtype: int64

In [77]:
## In percentage
country_group['SocialMedia'].value_counts(normalize=True).loc['Nepal']

SocialMedia
YouTube                     0.411255
Facebook                    0.307359
LinkedIn                    0.095238
Twitter                     0.069264
Reddit                      0.038961
Instagram                   0.034632
WhatsApp                    0.021645
I don't use social media    0.017316
WeChat 微信                   0.004329
Name: SocialMedia, dtype: float64

### Aggregate Function

In [78]:
country_group['Salary(USD)'].median().loc['Nepal']

4224.0

In [80]:
country_group['Salary(USD)'].agg(['mean', 'median']).loc['India']

mean      28057.664916
median    10080.000000
Name: India, dtype: float64

### Finding what % of respondent from certain country known python

In [83]:
filter_country = df['Country'] == 'India'
df.loc[filter_country]['LanguageWorkedWith']

Respondent
22013    Assembly;Bash/Shell/PowerShell;C;C++;Java;Java...
28243            C++;C#;Java;JavaScript;Objective-C;Python
72732                                      Java;Python;SQL
16793    Assembly;Bash/Shell/PowerShell;C;C++;C#;Dart;G...
52561                                        C#;JavaScript
                               ...                        
77339                                            Other(s):
79795                 Bash/Shell/PowerShell;Python;SQL;VBA
83862    Assembly;C;C++;HTML/CSS;Java;JavaScript;Object...
84299     HTML/CSS;Java;JavaScript;Python;Swift;TypeScript
86012        Bash/Shell/PowerShell;C++;HTML/CSS;JavaScript
Name: LanguageWorkedWith, Length: 9061, dtype: object

In [87]:
df.loc[filter_country]['LanguageWorkedWith'].str.contains('Python').sum()


## SO in india there are 3105people who knows python

3105

In [97]:
df['Country'].value_counts().loc['India']

9061

#### SO, out of 9061 respondents 3105 people know python

#### Doing same above things using group by

In [92]:
country_group['LanguageWorkedWith'].str.contains('Python').sum()

## Gives error

AttributeError: 'SeriesGroupBy' object has no attribute 'str'

In [93]:
country_group['LanguageWorkedWith'].apply(lambda x: x.str.contains('Python').sum())

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

### What % of people from each country know Python?

### 1st finding Total number of respondents of each country

In [98]:
country_respondents = df['Country'].value_counts()
country_respondents

United States                       20949
India                                9061
Germany                              5866
United Kingdom                       5737
Canada                               3395
                                    ...  
North Korea                             1
Niger                                   1
Saint Vincent and the Grenadines        1
Chad                                    1
Brunei Darussalam                       1
Name: Country, Length: 179, dtype: int64

### 2nd Finding total number of people who knows python

In [100]:
country_uses_python = country_group['LanguageWorkedWith'].apply(lambda x: x.str.contains('Python').sum())
country_uses_python

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

### Now concanating two series of dataframe

In [102]:
python_dataframe = pd.concat([country_respondents, country_uses_python], axis=1, sort=False)
python_dataframe

Country  LanguageWorkedWith
United States                       20949               10083
India                                9061                3105
Germany                              5866                2451
United Kingdom                       5737                2384
Canada                               3395                1558
...                                   ...                 ...
North Korea                             1                   0
Niger                                   1                   1
Saint Vincent and the Grenadines        1                   0
Chad                                    1                   0
Brunei Darussalam                       1                   0

[179 rows x 2 columns]

### Now Renaming columns

In [109]:
python_dataframe.rename(columns={'Country': 'TotalRespondents', 'LanguageWorkedWith': 'KnowsPython'}, inplace=True)

### Last Finding the % of people from each country who knoes python

In [110]:
python_dataframe.head()

TotalRespondents  KnowsPython
United States              20949        10083
India                       9061         3105
Germany                     5866         2451
United Kingdom              5737         2384
Canada                      3395         1558

In [111]:
python_dataframe['Total %'] = (python_dataframe['KnowsPython'] / python_dataframe['TotalRespondents'] * 100 )

In [112]:
python_dataframe

TotalRespondents  KnowsPython     Total %
United States                                20949        10083   48.131176
India                                         9061         3105   34.267741
Germany                                       5866         2451   41.783157
United Kingdom                                5737         2384   41.554820
Canada                                        3395         1558   45.891016
...                                            ...          ...         ...
North Korea                                      1            0    0.000000
Niger                                            1            1  100.000000
Saint Vincent and the Grenadines                 1            0    0.000000
Chad                                             1            0    0.000000
Brunei Darussalam                                1            0    0.000000

[179 rows x 3 columns]

In [114]:
python_dataframe.sort_values(by='Total %', ascending=False).head(20)

TotalRespondents  KnowsPython     Total %
Sao Tome and Principe                     1            1  100.000000
Niger                                     1            1  100.000000
Dominica                                  1            1  100.000000
Timor-Leste                               1            1  100.000000
Turkmenistan                              7            6   85.714286
Mauritania                                7            5   71.428571
Guinea                                    3            2   66.666667
Bahamas                                   3            2   66.666667
Guyana                                    3            2   66.666667
Uganda                                   72           47   65.277778
Iceland                                  49           30   61.224490
Namibia                                   5            3   60.000000
Haiti                                     5            3   60.000000
Benin                                     5            3   60.000000
Congo, Republic of the...                 7            4   57.142857
Oman                                     11            6   54.545455
Cuba                                     30           16   53.333333
Republic of Korea                        39           20   51.282051
Bhutan                                    2            1   50.000000
South Korea                             160           80   50.000000

In [115]:
python_dataframe.loc['Nepal']

TotalRespondents    237.000000
KnowsPython         101.000000
Total %              42.616034
Name: Nepal, dtype: float64

### Handeling Missing values (Cleaning Datatypes and Handeling Missing values)

In [120]:
df['YearsCode'].head(10)

Respondent
63129    Less than 1 year
50499                   7
39258                   2
58450                   6
7085                    7
22450                   3
48436                   2
10746                   4
8149                    2
29736                  10
Name: YearsCode, dtype: object

In [121]:
df['YearsCode'].unique()

array(['Less than 1 year', '7', '2', '6', '3', '4', '10', '36', '1', nan,
       'More than 50 years', '20', '9', '5', '8', '31', '28', '15', '13',
       '12', '17', '18', '21', '35', '14', '25', '11', '19', '22', '23',
       '40', '39', '30', '43', '27', '33', '29', '24', '34', '16', '32',
       '26', '38', '44', '41', '45', '37', '47', '42', '46', '48', '50',
       '49'], dtype=object)

In [122]:
df['YearsCode'].nunique()

52

In [123]:
df['YearsCode'].replace('Less than 1 year', 0, inplace=True)
df['YearsCode'].replace('More than 50 years', 51, inplace=True)

In [124]:
df['YearsCode'].unique()

array([0, '7', '2', '6', '3', '4', '10', '36', '1', nan, 51, '20', '9',
       '5', '8', '31', '28', '15', '13', '12', '17', '18', '21', '35',
       '14', '25', '11', '19', '22', '23', '40', '39', '30', '43', '27',
       '33', '29', '24', '34', '16', '32', '26', '38', '44', '41', '45',
       '37', '47', '42', '46', '48', '50', '49'], dtype=object)

#### changing dtypes 

In [125]:
df['YearsCode'] = df['YearsCode'].astype(float)

In [126]:
df['YearsCode'].mean()

11.662114216834588

In [127]:
df['YearsCode'].median()

9.0